quadsim/robots/motor_model.py

```diff
@@ -0,0 +1,104 @@
+"""Implements a simple DC motor model."""
+import enum
+import numpy as np
+from typing import Sequence, Union, Optional
+
+POSITION_INDEX = 0
+KP_INDEX = 1
+VELOCITY_INDEX = 2
+KD_INDEX = 3
+TORQUE_INDEX = 4
+COMMAND_DIMENSION = 5
+
+
+class MotorControlMode(enum.Enum):
+    """Different motor control modes.
+    See the documentation of MotorModel class for details."""
+
+    POSITION = 0
+    TORQUE = 1
+    HYBRID = 2
+
+
+class MotorModel:
+    """Implements a simple DC motor model for simulation.
+
+    To accurately model the motor behaviors, this class converts all motor
+    commands into torques, which could be send directly to the simulator.
+    Right now, 3 motor control modes are supported:
+    - POSITION: performs joint-level PD control
+    - TORQUE: directly takes in motor torque command
+    - HYBRID: takes in a 5-d tuple (pos, kp, vel, kd, torque), and output
+      torque is a sum of PD torque and additional torque.
+    """
+
+    def __init__(self, config) -> None:
+        """Initializes the class."""
+        self.config = config
+        self._motor_control_mode = self.config.motor_control_mode
+        self._num_motors = self.config.num_motors
+        self._kp = self.config.kps
+        self._kd = self.config.kds
+        self._strength_ratios = np.ones(self._num_motors)
+        # Thresholds
+        self._max_torque = self.config.max_torque
+        self._min_torque = self.config.min_torque
+        self._max_velocity = self.config.max_velocity
+        self._min_velocity = self.config.min_velocity
+        self._max_position = self.config.max_position
+        self._min_position = self.config.min_position
+
+    def set_motor_gains(
+        self, kp: Union[float, Sequence[float]], kd: Union[float, Sequence[float]]
```

Comment:
I actually like `Sequence` much better than `List` or `Tuple` explicitly. I'm gonna use it from now on 🙂 . I am wondering about whether its cleaner to do:
1. `Union[float, Sequence[float]]` or
2. `Sequence[float, ...]`

I guess with the latter, you always need to index into a datastructure, whereas with the former, sometimes you have a non-nested datastructure.

Reply:
The reason I implemented it this way is to allow the function to take in either a single float or a list of 12 floats (the `np.ones()*` stuff takes care of this). I agree it's not that clear here, maybe we should make it more straightforward?